### DISCUSSION TITLE
openssl 3.13  sm2 sign in EVP_DigestSignUpdate   get  Segmentation fault

### DISCUSSION THREAD

**Discussion by alanmoleapfive** [openssl/openssl#22931] [Q&A]
bool sm2_sign(EVP_PKEY* prikey, const unsigned char* data, size_t data_len, unsigned char* sign, size_t* sign_len) {
    unsigned char sm2_id[]  = "411715975@qq.com";
    unsigned int sm2_id_len = sizeof(sm2_id);
    if (!prikey || !data || data_len == 0 || !sign || !sign_len) {
        std::cerr << "Invalid input parameters" << std::endl;
        return false;
    }
    EVP_MD_CTX* mctx = EVP_MD_CTX_new();
    if (!mctx) {
        std::cerr << "Error creating EVP_MD_CTX" << ERR_error_string(ERR_get_error(), nullptr)<< std::endl;
        return false;
    }
    EVP_PKEY_CTX* pctx = EVP_PKEY_CTX_new(prikey, NULL);
    if (!pctx) {
        std::cerr << "Error creating EVP_PKEY_CTX_new" << ERR_error_string(ERR_get_error(), nullptr)<< std::endl;
         EVP_MD_CTX_free(mctx);
        return false;
    }
    if (EVP_PKEY_CTX_set1_id(pctx, sm2_id, sm2_id_len) <= 0) {
        std::cerr << "Error creating EVP_PKEY_CTX_set1_id" << ERR_error_string(ERR_get_error(), nullptr)<< std::endl;
        EVP_MD_CTX_free(mctx);
        EVP_PKEY_CTX_free(pctx);
        return false;
    }
    
    EVP_MD_CTX_set_pkey_ctx(mctx, pctx);

    std::cout << "Data: ";
    for (size_t i = 0; i < data_len; ++i) {
        std::cout << static_cast<char>(data[i]);
    }
    std::cout << std::endl;
    std::cout << "Data length: " << data_len << std::endl;

    std::cout << "EVP_DigestSignInit" << std::endl;
    if (EVP_DigestSignInit(mctx, nullptr, EVP_sm3(), nullptr, prikey) != 1) {
        std::cerr << "Error initializing EVP_DigestSignInit" << ERR_error_string(ERR_get_error(), nullptr)<< std::endl;
        EVP_MD_CTX_free(mctx);
        EVP_PKEY_CTX_free(pctx);
        return false;
    }
    std::cout << "EVP_DigestSignUpdate" << std::endl;
    if (EVP_DigestSignUpdate(mctx, data, data_len) != 1) {
        std::cerr << "Error updating EVP_DigestSignUpdate" << ERR_error_string(ERR_get_error(), nullptr)<< std::endl;
        EVP_MD_CTX_free(mctx);
        EVP_PKEY_CTX_free(pctx);
        return false;
    }
    std::cout << "EVP_DigestSignFinal" << std::endl;
    if (EVP_DigestSignFinal(mctx, sign, sign_len) != 1) {
        std::cerr << "Error finalizing EVP_DigestSignFinal" << ERR_error_string(ERR_get_error(), nullptr)<< std::endl;
        EVP_MD_CTX_free(mctx);
        EVP_PKEY_CTX_free(pctx);
        return false;
    }

    EVP_MD_CTX_free(mctx);
    EVP_PKEY_CTX_free(pctx);
    return true;
}

Data: 123456
Data length: 7
EVP_DigestSignInit
EVP_DigestSignUpdate
Segmentation fault

**Comment by andylee521** [comment]
I think you'd better like this:（Pseudocode）
```
EVP_MD_CTX* mctx = EVP_MD_CTX_new();
EVP_PKEY_CTX* pctx = EVP_PKEY_CTX_new_from_pkey(testctx, pkey, testpropq)；
EVP_DigestSignInit(mctx , NULL, EVP_sm3(), NULL, pkey);
EVP_PKEY_CTX_set1_id(pctx, sm2_id, sizeof(sm2_id));
EVP_DigestSignUpdate(mctx , data, data_len);
EVP_DigestSignFinal(mctx, NULL, &sign_len);
sign = OPENSSL_malloc(sign_len);
EVP_DigestSignFinal(md_ctx, sig, &sig_len);
```
